src/pymor/algorithms/rand_la.py

```diff
@@ -187,75 +371,87 @@ def random_generalized_svd(A, range_product=None, source_product=None, modes=6,
+        Source product |Operator| :math:`T` w.r.t. which the randomized SVD is computed.
+    subspace_iterations
+        The number of subspace iterations to increase the relative weight
+        of the larger singular values (ignored when `single_pass` is `True`).
+    oversampling
+        The number of samples that are drawn in addition to the desired basis size in the
+        randomized range approximation process.
 
     Returns
     -------
     U
-        |VectorArray| of approximated left singular vectors.
+        |VectorArray| containing the approximated left singular vectors.
     s
         One-dimensional |NumPy array| of the approximated singular values.
-    Vh
-        |VectorArray| of the approximated right singular vectors.
+    V
+        |VectorArray| containing the approximated right singular vectors.
     """
-    logger = getLogger('pymor.algorithms.rand_la')
+    logger = getLogger('pymor.algorithms.rand_la.randomized_svd')
 
-    assert isinstance(A, Operator)
+    RRF = RandomizedRangeFinder(A, subspace_iterations=subspace_iterations, range_product=range_product,
+                                source_product=source_product)
 
-    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
-    assert 0 <= p <= max(A.source.dim, A.range.dim) - modes and isinstance(p, int)
-    assert q >= 0 and isinstance(q, int)
+    assert 0 <= n <= max(A.source.dim, A.range.dim) and isinstance(n, int)
+    assert 0 <= oversampling and isinstance(oversampling, int)
+    if oversampling > max(A.source.dim, A.range.dim) - n:
+        logger.warning('Oversampling parameter is too large!')
+        oversampling = max(A.source.dim, A.range.dim) - n
+        logger.info(f'Setting oversampling to {oversampling} and proceeding ...')
 
     if range_product is None:
         range_product = IdentityOperator(A.range)
-    else:
-        assert isinstance(range_product, Operator)
-        assert range_product.source == range_product.range == A.range
-
     if source_product is None:
         source_product = IdentityOperator(A.source)
-    else:
-        assert isinstance(source_product, Operator)
-        assert source_product.source == source_product.range == A.source
-
     if A.source.dim == 0 or A.range.dim == 0:
         return A.source.empty(), np.array([]), A.range.empty()
 
-    Q = rrf(A, source_product=source_product, range_product=range_product, q=q, l=modes+p)
-    B = A.apply_adjoint(range_product.apply(Q))
-    Q_B, R_B = gram_schmidt(source_product.apply_inverse(B), product=source_product, return_R=True)
-    U_b, s, Vh_b = sp.linalg.svd(R_B.T, full_matrices=False)
+    with logger.block('Approximating basis for the operator range ...'):
+        Q = RRF.find_range(basis_size=n+oversampling)
+
+    with logger.block('Projecting operator onto the reduced space ...'):
+        if isinstance(source_product, IdentityOperator):
+            R_B = A.apply_adjoint(Q).to_numpy().T
```

Comment:
Here you assume that `A.apply_adjoint(Q)` can be easily converted into a numpy array. We easily make that assumption in pyMOR ...

In view of earlier discussions: in the future we could call here pyMOR's `va_svd` or even `svd` method (I still think we should not mix both cases) which then can decide to dispatch into a direct solver which does not first perform a QR decomposition.

Reply:
Okay so I take it, you suggest to leave it like this for now?

Reply:
I have now changed the implementation to use `qr_svd`. This should also make it easier in the future to drop in a dispatcher. I did some short tests between the two versions. If products are given, `qr_svd` is a tad faster. It falls behind for larger orders (200 and above) but it is acceptable (about factor 2 for n=800).